jax/_src/numpy/fft.py

```diff
@@ -55,7 +58,20 @@ def _fft_core(func_name, fft_type, a, s, axes, norm):
     axes = tuple(range(a.ndim - len(axes), a.ndim))
     a = jnp.moveaxis(a, orig_axes, axes)
 
-  if s is None:
+  if s is not None:
+    a = jnp.asarray(a)
+    in_s = list(a.shape)
+    for axis, x in zip(axes, s):
+      in_s[axis] = x
+    if fft_type == xla_client.FftType.IRFFT:
+      in_s[-1] = (in_s[-1] // 2 + 1)
+
+    if any([x != y for x, y in zip(in_s, a.shape)]):
```

Comment:
Nit: this optimization probably isn't needed, as long as fft is called inside jit compiled code. I would remove it for now.

Reply:
I will removed this. I can see with 0 padding that this will be included in the XLA code, but I am not familiar enough with the internals to know if this matters or not.
```python
import jax
import numpy as np
import jax.numpy as jnp
def f(sig):
    return jnp.pad(sig, [(0, 0)])
print(jax.xla_computation(f)(jnp.asarray(np.random.randn(5))).as_hlo_text())
```

<details>
<summary>Output</summary>


```
HloModule xla_computation_f__1.64

jit__pad__1.6 {
  constant.11 = pred[] constant(false)
  parameter.9 = f32[5]{0} parameter(2)
  parameter.10 = s32[] parameter(3)
  broadcast.12 = s32[1,2]{1,0} broadcast(parameter.10), dimensions={}
  convert.13 = f32[1,2]{1,0} convert(broadcast.12)
  parameter.7 = s32[0]{0} parameter(0)
  convert.23 = s32[0]{0} convert(parameter.7)
  constant.14 = s32[] constant(0)
  constant.15 = s32[] constant(0)
  compare.16 = pred[] compare(constant.14, constant.15), direction=LT
  constant.17 = s32[] constant(0)
  constant.18 = s32[] constant(1)
  add.19 = s32[] add(constant.17, constant.18)
  constant.20 = s32[] constant(0)
  select.21 = s32[] select(compare.16, add.19, constant.20)
  broadcast.22 = s32[1]{0} broadcast(select.21), dimensions={}
  concatenate.24 = s32[1]{0} concatenate(convert.23, broadcast.22), dimensions={0}
  constant.25 = s32[] constant(0)
  constant.26 = s32[] constant(0)
  compare.27 = pred[] compare(constant.25, constant.26), direction=LT
  constant.28 = s32[] constant(0)
  constant.29 = s32[] constant(2)
  add.30 = s32[] add(constant.28, constant.29)
  constant.31 = s32[] constant(0)
  select.32 = s32[] select(compare.27, add.30, constant.31)
  broadcast.33 = s32[1]{0} broadcast(select.32), dimensions={}
  concatenate.34 = s32[2]{0} concatenate(concatenate.24, broadcast.33), dimensions={0}
  gather.35 = f32[] gather(convert.13, concatenate.34), offset_dims={}, collapsed_slice_dims={0,1}, start_index_map={0,1}, index_vector_dim=0, slice_sizes={1,1}
  pad.36 = f32[5]{0} pad(parameter.9, gather.35), padding=0_0
  parameter.8 = s32[0]{0} parameter(1)
  convert.46 = s32[0]{0} convert(parameter.8)
  constant.37 = s32[] constant(0)
  constant.38 = s32[] constant(0)
  compare.39 = pred[] compare(constant.37, constant.38), direction=LT
  constant.40 = s32[] constant(0)
  constant.41 = s32[] constant(1)
  add.42 = s32[] add(constant.40, constant.41)
  constant.43 = s32[] constant(0)
  select.44 = s32[] select(compare.39, add.42, constant.43)
  broadcast.45 = s32[1]{0} broadcast(select.44), dimensions={}
  concatenate.47 = s32[1]{0} concatenate(convert.46, broadcast.45), dimensions={0}
  constant.48 = s32[] constant(1)
  constant.49 = s32[] constant(0)
  compare.50 = pred[] compare(constant.48, constant.49), direction=LT
  constant.51 = s32[] constant(1)
  constant.52 = s32[] constant(2)
  add.53 = s32[] add(constant.51, constant.52)
  constant.54 = s32[] constant(1)
  select.55 = s32[] select(compare.50, add.53, constant.54)
  broadcast.56 = s32[1]{0} broadcast(select.55), dimensions={}
  concatenate.57 = s32[2]{0} concatenate(concatenate.47, broadcast.56), dimensions={0}
  gather.58 = f32[] gather(convert.13, concatenate.57), offset_dims={}, collapsed_slice_dims={0,1}, start_index_map={0,1}, index_vector_dim=0, slice_sizes={1,1}
  pad.59 = f32[5]{0} pad(pad.36, gather.58), padding=0_0
  ROOT tuple.60 = (f32[5]{0}) tuple(pad.59)
}

ENTRY xla_computation_f__1.64 {
  constant.4 = pred[] constant(false)
  constant.1 = s32[0]{0} constant({})
  constant.2 = s32[0]{0} constant({})
  parameter.3 = f32[5]{0} parameter(0)
  constant.5 = s32[] constant(0)
  call.61 = (f32[5]{0}) call(constant.1, constant.2, parameter.3, constant.5), to_apply=jit__pad__1.6
  get-tuple-element.62 = f32[5]{0} get-tuple-element(call.61), index=0
  ROOT tuple.63 = (f32[5]{0}) tuple(get-tuple-element.62)
}
```
</details>